cargo-shuttle/src/init.rs

```diff
@@ -1,56 +1,291 @@
+        Ok(())
+    })?;
+
+    // Initialize a Git repository in the destination directory if there
+    // is no existing Git repository present in the surrounding folders.
+    let no_git_repo = gix::discover(&dest).is_err();
+    if no_git_repo {
+        gix::init(&dest).context("Failed to initialize project repository")?;
+    }
+
+    Ok(())
+}
+
+/// Location where the template to generate a project is found.
+#[derive(Debug, Clone)]
+enum GenerateFrom {
+    /// `LocalPath` also supports sub-folders, but it doesn't need
+    /// multiple entries for this. The primary path and the sub-folder
+    /// path are simply concatenated when the `LocalPath` is constructed.
+    LocalPath(PathBuf),
+    Url {
+        url: String,                // e.g. `https://github.com/shuttle-hq/shuttle
+        subfolder: Option<PathBuf>, // e.g. `cargo-shuttle`
+    },
+    RemoteRepo {
+        vendor: GitVendor,
+        owner: String,              // e.g. `shuttle-hq`
+        name: String,               // e.g. `shuttle`
+        subfolder: Option<PathBuf>, // e.g. `cargo-shuttle`
+    },
+}
+
+impl GenerateFrom {
+    // Very loose restrictions are applied to repository names.
+    // What's important is that all names that are valid by the vendor's
+    // rules are accepted here. There is no need to check that the user
+    // actually provided a name that the vendor would accept.
+    const GIT_PATTERN: &str = "^(?:(?<vendor>gh|gl|bb):)?(?<owner>[^/.:]+)/(?<name>[^/.:]+)$";
+
+    fn from(loc: TemplateLocation) -> Result<Self> {
+        let git_re = Regex::new(Self::GIT_PATTERN).unwrap();
+
+        if let Some(caps) = git_re.captures(&loc.auto_path) {
+            let vendor = match caps.name("vendor").map(|v| v.as_str()) {
+                Some("gl") => GitVendor::GitLab,
+                Some("bb") => GitVendor::BitBucket,
+                // GitHub is the default vendor if no other vendor is specified.
+                Some("gh") | None => GitVendor::GitHub,
+                Some(_) => unreachable!("should never match unknown vendor"),
+            };
+
+            Ok(Self::RemoteRepo {
+                vendor,
+                // `owner` and `name` are required for the regex to
+                // match. Thus, we don't need to check if they exist.
+                owner: caps["owner"].to_owned(),
+                name: caps["name"].to_owned(),
+                subfolder: loc.subfolder.map(PathBuf::from),
+            })
+        } else if Path::new(&loc.auto_path).is_absolute() || loc.auto_path.starts_with('.') {
+            // Local paths to template locations must be absolute or start with a
+            // dot pattern. This way, conflicts between relative path in the form
+            // `foo/bar` and repository name in the form `owner/name` are avoided.
+            match loc.subfolder {
+                Some(subfolder) => Ok(Self::LocalPath(Path::new(&loc.auto_path).join(subfolder))),
```

Comment:
Just a thought: if we are to add branch/rev selection to the clone operation, we would need to require a local path to be handled like the other alternatives, so that the clone and checkout can be done correctly.

Reply:
Do you mean in cases where there is a Git repository at the local path? Yes, that's a good idea.

The `generate` function ought to be divided into a *setup* which ends in the same generic state, no matter what the initial form of the origin was. This state could be the raw template places in a temporary directory. After the *setup*, the rest of the code would be the same for local folders and repository clones. This code always uses `subfolder`.

I actually prefer this design over the current one. It separates the different stages and the information required at each one of them much better.

Reply:
Yeah it's a bit more streamlined.
I wonder if the `depth 1` and potential branch/rev parameter clash in some way :thinking: